Adding dev installation guide

markusjura opened the pull request:
This PR is not only adding documentation for ConductR sandbox but based also refactors the entire developer guide.

The changes should not be merged until the new version of sbt-conductr-sandbox and conductr has been released incl. the PRs:
- https://github.com/typesafehub/sbt-conductr-sandbox/pull/13
- https://github.com/typesafehub/conductr/pull/785

**Review**
Because a lot of files has been touched probably the best approach to review the changes is to run this branch locally.

**Why was it necessary to refactor the developers guide**
ConductR's documentation aims to be cohesive throughout the entire development guide (at least until 'Deploying bundle'. Now the section 'Creating bundle' introduces `sbt-bundle`. Afterwards the section 'Deploying bundle' introduces `sbt-conductr` which includes `sbt-bundle`. If we would have just added a section afterwards which introduces `sbt-conductr-sandbox` and also includes `sbt-bundle` and can be used alongside `sbt-conductr` it wouldn't have been the best first user experience.

Now the developers guide starts with a `Quick start` which will cover all the aspects of configuring an application, creating and deploying a bundle and start a ConductR cluster briefly. This quick starts also references to the more detailed sections which describes the certain topics in more detail.

The installation section in the `Getting started` group contains now an additional item how to install ConductR for development purposes. This chapter only explains that a Docker image exists for this purpose and then links to the developers quick start guide.

markusjura committed:
Adding dev installation guide

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
```diff
@@ -0,0 +1,70 @@
+# ConductR sandbox cluster
```
Remove the word "cluster" - the emphasis is just on managing ConductR. In fact, we're only providing the single node version for development; we may provide a cluster-able version in the future, but not right now.

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
```diff
@@ -0,0 +1,70 @@
+# ConductR sandbox cluster
+
+In order to manage a ConductR cluster we provide a sbt plugin [sbt-conductr-sandbox](https://github.com/typesafehub/sbt-conductr-sandbox). Its purpose is to simplify the setup of a ConductR cluster during development or on a continous integration platform.  By default it uses the conductr-dev Docker image which gets automatically pulled down from Typesafe's public docker repository on bintray whe ...
```
Remove "cluster" from here too. Perhaps start with, "In order to support development activities around ConductR we provide...".

huntc replied:
In particular, the initial use-case is that a developer needs to be able to test their bundle's configuration prior to deploying into a production environment.

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
````diff
@@ -0,0 +1,70 @@
+[info] Running container cond-0 exposing 192.168.59.103:9000 192.168.59.103:9909...
+```
+
+The `visualization` feature provides a web interface to visualize the ConductR cluster together with deployed and running bundles. After enabling the feature, access it at http://docker-host-ip:9909. For convience, the url of the visualizer app is displayed in the sbt session, e.g. http://192.168.59.103:9909.
````
Perhaps emphasise that the `docker-host-ip` should be substituted:

```
http://{docker-host-ip}:9909
```

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
````diff
@@ -0,0 +1,70 @@
+The `logging` feature consolidates the logging output of ConductR itself and the bundles that it executes. To view the consolidated log messsages enable [sbt-conductr](https://github.com/sbt/sbt-conductr) and then run:
+
+```scala
+conduct logs
````
A bundle id arg is required. For example: `conduct logs conductr-elasticsearch` in order to see the ES logs.

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
````diff
@@ -0,0 +1,70 @@
+```scala
+sandbox stop
+```
+
````
I think a section on debugging would be super useful here.

huntc commented on src/main/play-doc/developer/DevIntro.md:
```diff
@@ -1,9 +0,0 @@
-# Developer introduction
```
I still think that this introduction is somewhat important. It sets the scene. I felt that it was particularly important to emphasise that we're big on avoiding vendor-lockin.

huntc commented on src/main/play-doc/developer/DevQuickStart.md:
```diff
@@ -0,0 +1,129 @@
+
+[[images/visualizer_with_app.png]]
+
+That's it! You now have ConductR running with the visualizer and your own application.
```
Nice

huntc commented on src/main/play-doc/developer/ConductrSandboxCluster.md:
```diff
@@ -0,0 +1,70 @@
+
+#### ConductR features
+
+ConductR gets shipped with features which can be optionally enabled during startup by specifying the `--withFeatures` option, e.g.:
```
It is important to mention how, if you're using boot2docker, then you'll need to increase its memory size to at least 2GB. Here's how, at least on the Mac:

```
boot2docker stop
VBoxManage modifyvm boot2docker-vm --memory 2048 
boot2docker start
```

huntc commented on src/main/play-doc/gettingStarted/Install.md:
```diff
@@ -11,14 +12,22 @@ If you want to first get familiar with the terms and components of ConductR go t
 * [[ConductR components|ConductrComponents]]
 
 
+# Dev Installation
```
I think that mixing in developer concerns into this section should be avoided. We should be mindful of the two audiences for our doc. This doc is particularly focused on ops.

markusjura commented:
Changes made as discussed:

**ConductR sandbox**
- Removed cluster
- Added 'Debugging ConductR' section
- Fixed command to log messages

**Getting started restructuring**
- Removed getting started guide
- Instead added chapter in Home to navigate user to either the operations or developer guide
- Moved Installation section to operations guide

**Misc**
- Changed headlines to plural names

@huntc Ready for another review.

huntc commented on src/main/play-doc/developer/ConductrSandbox.md:
````diff
@@ -0,0 +1,110 @@
+    ```scala
+    sandbox debug
+    ```
+2. This exposes a debug port to docker and the ConductR container. By default the debug port is set to `5005` which is the default port for remote debugging in IntelliJ. If you are happy with the default setting skip this step. Otherwise specify another port in the `build.sbt`:
````
"for remote debugging in IDEs such as IntelliJ"

markusjura replied:
Corrected.

huntc commented on src/main/play-doc/developer/ConductrSandbox.md:
```diff
@@ -0,0 +1,110 @@
+    - Port: `debugPort` of your application, e.g. 5005
+
+    [[images/conductr_sandbox_debugging_intellij.png]]
+3. Save this configuration and start remote debugging with `Run` -> `Debug` and then select your configuration.
```
If we're going to have a section for IntelliJ then I think we should have one for Eclipse... WDYT?

huntc replied:
Otherwise may be just remove the IntelliJ section.

markusjura replied:
Eclipse description has been added.

huntc commented:
LGTM

markusjura commented:
PR #53 has been incorporated.
